File: scala/103.scala
import scala.collection.mutable.Queue
import scala.collection.mutable.ListBuffer
object Solution {
    def zigzagLevelOrder(root: TreeNode): List[List[Int]] = {
        val q = Queue[TreeNode]()
        if (root != null)
          q.enqueue(root)
        val res = ListBuffer[List[Int]]()
        var flip = false
        while (!q.isEmpty) {
            val s = q.size
            val level = ListBuffer[Int]()
            for (i <- 0 until s) {
                val n = q.dequeue()
                level += n.value
                if (n.left != null) q.enqueue(n.left)
                if (n.right != null) q.enqueue(n.right)
            }
            if (flip)
              res += level.toList.reverse
            else 
              res += level.toList
            flip = !flip
        }
        res.toList
    }
}

File: leetcodeJava/103.java
package leetcodeJava;

class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        if (root == null) return new LinkedList();
        var q = new LinkedList<TreeNode>();
        q.add(root);
        var res = new LinkedList<List<Integer>>();
        var flag = false;
        while (q.size() != 0) {
            var level = new LinkedList<Integer>();
            var size = q.size();
            for (var i = 0; i < size; i++) {
                var node = q.pop();
                level.add(node.val);
                if (node.left != null) q.add(node.left);
                if (node.right != null) q.add(node.right);
            }
            if (flag) Collections.reverse(level);
            flag = !flag;
            res.add(level);
        }
        return res;
    }
}
